File: cpp/316_去除重复字母_medium.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        vector<char> flag(26, 0), num(26, 0);
        for(char ch : s)
            num[ch-'a']++;
        string stack;
        for(char ch : s){
            if(!flag[ch-'a']){
                while(!stack.empty() && stack.back()>ch){
                    if(num[stack.back()-'a'] > 0){
                        flag[stack.back()-'a'] = 0;
                        stack.pop_back();
                    }
                    else
                        break;
                }
                stack.push_back(ch);
                flag[ch-'a'] = 1;
            }
            num[ch-'a']--;
        }
        return stack;
    }
};

File: python/316_去除重复字母_medium.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        flag, nums = [0 for i in range(26)], [0 for j in range(26)]
        stack = []
        for ch in s:
            nums[ord(ch)-ord('a')] += 1
        for ch in s:
            if not flag[ord(ch)-ord('a')]:
                while stack and stack[-1] > ch:
                    if nums[ord(stack[-1]) - ord('a')] > 0:
                        flag[ord(stack[-1]) - ord('a')] = 0
                        stack.pop()
                    else:
                        break
                stack.append(ch)
                flag[ord(stack[-1]) - ord('a')] = 1
            nums[ord(ch) - ord('a')] -= 1
        return ''.join(stack)